ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -575,19 +621,68 @@ KEXT_STATIC int HandleVnodeOperation(
             
             if (ActionBitIsSet(action, KAUTH_VNODE_WRITE_DATA | KAUTH_VNODE_APPEND_DATA))
             {
-                if (!TryGetVirtualizationRoot(&perfTracer, context, currentVnode, pid, CallbackPolicy_UserInitiatedOnly, &root, &vnodeFsidInode, &kauthResult, kauthError))
+                /* At this stage we know the file is NOT empty, but it could be
+                 * in either placeholder (hydrated) or full (modified/added to
+                 * watchlist) state.
+                 *
+                 * If it's a placeholder, we need to notify the provider before
+                 * allowing modifications so that it is tracked as a full file
+                 * (and the placeholder xattr is removed).
+                 * If the provider is offline in this situation, we want to
+                 * prevent writes (by most processes) to the file so that the
+                 * modifications don't go undetected.
+                 *
+                 * If it's already a full file on the other hand, nothing needs
+                 * to be done, so we can allow writes even when the provider is
+                 * offline.
+                 */
+                ProviderStatus providerStatus;
+                bool shouldMessageProvider = TryGetVirtualizationRoot(
```

Comment:
nit: do we need a seperate varaible for shouldMessageProvider?

Reply:
> Isn't there also the case where we can have two threads do
> 
> resizeArrayLength = MIN(MAX(1u, s_maxOfflineIOPIDs * 2u), UINT32_MAX);
> And allocate a block of memory for that size?
> 
> If so I think everything works we just copy twice. Perhaps a comments explaining that to make it simple for the reader.

No, we shouldn't copy twice as the race will be detected by the `resizeArrayLength > s_maxOfflineIOPIDs` check, in which case one thread's alloc'd memory is freed again without using it. At least that's what the code is *supposed* to do, please check that also matches your interpretation. :-)

Reply:
I found it more readable this way, given the 12-line `TryGetVirtualizationRoot` function call.

Reply:
I get that.  All of them are 12-lines though.  Since the other calls don't assign a variable and do an assert check later on I'd prefer to keep it consistent.  Otherwise me in a year reading this code goes, why did we do this one differently than the others?   

If you feel strongly about it can you add a comment as to why it breaks the convention the other ones have?  It certainly isn't wrong, but it will be a helpful refresher on why it's different.

Reply:
I don't like the other call sites either, but the logic that follows them is much less complex so I don't feel the meaning of everything needs spelling out so much. I don't mind removing the variable though. (and the `assert()`s that check it later on)